sycl/doc/extensions/proposed/sycl_ext_oneapi_async_memory_alloc.asciidoc

```diff
@@ -0,0 +1,755 @@
+  to the source. The default deallocation threshold is zero.
+* No memory is immediately allocated on construction, instead memory is
+  allocated from the source for the pool when requested via the asynchronous
+  malloc and free functions.
+* Memory is freed back to the source when the SYCL queue which enqueued the
+  respective allocations and frees is synchronized with, i.e. when `queue::wait`
+  or `queue::wait_and_throw` is called, unless the current total memory
+  allocated to the memory pool under the deallocation threshold, in which case
+  the memory may be retained by the pool, though this can vary depending on
+  implementation defined parameters.
+* Memory is allocated as USM memory, in one of the USM memory allocation kinds
+  enumerated in `usm::alloc`, this is specified on construction of the
+  `memory_pool` object.
+* They are associated with a specific context and one or more device(s),
+  depending on the allocation kind, this is specified on construction of the
+  `memory_pool` object.
+
+Memory pools are intended to be used for both in-order and out-of-order SYCL
+queues.
+
+[source,c++]
+----
+namespace ext::oneapi::experimental {
+
+class memory_pool {
+
+  template <typename Properties = empty_properties_t>
+  memory_pool(context ctx, Properties props = {});
+
+  template <typename Properties = empty_properties_t>
+  memory_pool(context ctx, device dev, usm::alloc kind, Properties props = {});
+
+  ~memory_pool();
+
+  context get_context() const;
+
+  std::vector<device> get_devices() const;
+
+  usm::alloc get_alloc_kind() const;
+
+  size_t get_max_size() const;
+
+  size_t get_threshold() const;
+
+  void set_new_threshold(size_t newThreshold);
+
+}; // memory_pool
+
+}  // ext::oneapi::experimental
+----
+
+[source, c++]
+----
+template <typename Properties = empty_properties_t>
+memory_pool(context ctx, Properties props = {});
+----
+
+_Effects_: Constructs a memory pool associated with `ctx` and all SYCL devices
+associated with it, with the allocation kind `usm::host` and applying any
+properties in `props`.
+
+[source, c++]
+----
+template <typename Properties = empty_properties_t>
+memory_pool(context ctx, device dev, usm::alloc kind, Properties props = {});
+----
+
+_Effects_: Constructs a memory pool associated with `ctx` and `dev`, with the
+allocation kind `kind` and applying any properties in `props`.
+
+_Throws_: An exception with the `errc::invalid` error code if `kind` is
+`usm::host`.
+
+[source, c++]
+----
+~memory_pool();
+----
+
+_Effects_: If this was the last copy, signals to the SYCL runtime for the memory
+pool to be destroyed after all remaining allocations have been freed, and
+returns immediately without waiting.
+
+[source, c++]
+----
+context get_context() const;
+----
+
+_Returns_: The SYCL context associated with the memory pool.
+
+[source, c++]
+----
+std::vector<device> get_devices() const;
+----
+
+_Returns_: The SYCL device(s) associated with the memory pool.
```

Comment:
So this depends on how the memory pol was created, if it is created for host allocations it will return all devices associated with the context, however, if it is created for device or shared allocations it will return the specific device that was specified during construction only.

Reply:
This is not clear from the description, but I'm not sure this is the right API.  Do we really want to return a vector of devices here?  It seems like applications will really only care about the case when the pool contains "shared" or "device" memory, and there is just one device in those cases.

For comparison, consider the function `sycl::get_pointer_device`, which returns just a single device:

```
device get_pointer_device(const void* ptr, const context& syclContext)
```

I think we could do something similar for the memory pool.  I see two options:

* For "shared" and "device" USM pools, return the device that is specified in the pool constructor.  For "host" USM pools, arbitrarily return the first device in the context (same as `get_pointer_device`), or
* For "shared" and "device" USM pools, return the device that is specified in the pool constructor.  For "host" USM pools, throw an exception.

Reply:
I think I agree, my rationale for the current design was to return all devices, but this does mean creating a vector of devices so we should adopt one of the alternatives you suggested.

I would prefer to match the behaviour of `get_pointer_device` for consistency, I find it a little odd to return just the first device which is associated with the context, but in this case since the memory pool is associated with the whole context, the devices is not particularly relevant, so I think this is acceptable.